xcm/src/v1/multilocation.rs

```diff
@@ -329,18 +329,40 @@ impl MultiLocation {
 	///
 	/// Does not modify `self` in case of overflow.
 	pub fn reanchor(&mut self, target: &MultiLocation, ancestry: &MultiLocation) -> Result<(), ()> {
-		// TODO: https://github.com/paritytech/polkadot/issues/4489 Optimize this.
-
-		// 1. Use our `ancestry` to figure out how the `target` would address us.
-		let inverted_target = ancestry.inverted(target)?;
-
-		// 2. Prepend `inverted_target` to `self` to get self's location from the perspective of
-		// `target`.
-		self.prepend_with(inverted_target).map_err(|_| ())?;
+		if target.parents == self.parents {
+			// Find the first junction that is different as we travel along the paths from ancestry to self and target
+			let mut first_different = None;
+			for i in 0..target.interior.len() {
+				if target.interior.at(i) != self.interior.at(i) {
+					first_different = Some(i);
+					break
+				}
+			}
 
-		// 3. Given that we know some of `target` ancestry, ensure that any parents in `self` are
-		// strictly needed.
-		self.simplify(target.interior());
+			if let Some(i) = first_different {
+				// In the case where the self and target paths split, remove just the shared path
+				self.parents = (target.interior.len() - i) as u8;
+				for _ in 0..i {
+					self.take_first_interior();
+				}
+			} else {
+				// In the case where the paths are the same all the way to the target,
+				// remove all the junctions on the shared path
+				for _ in 0..target.interior.len() {
+					self.parents = 0;
+					self.take_first_interior();
+				}
+			}
+		} else if target.parents > self.parents {
+			// Handle when the target has more parents so we have to go through ancestry
+			self.parents = target.len() as u8 - 1;
+			self.interior
+				.push_front(ancestry.interior.first().unwrap_or(&Junction::OnlyChild).clone())
```

Comment:
This is only grabbing the first junction off of the ancestry? That doesn't seem correct to me. Could you try creating test cases where the `ancestry` isn't just `Parachain(2000)`? For example, imagine we're on a smart contract, and so the `ancestry` could look something like `Parachain(2022)/PalletInstance(50)/GeneralIndex(32)`.

Reply:
Good point, I made a false assumption that the ancestry only had once junction, which made it hard to reason about the multiple parents as well. I'm confused at what the multilocation ancestry is relative to - will ancestry ever have parents > 0? Are `self` and `target` MultiLocations relative to the first junction in `ancestry` or the last one?

Reply:
Also, I'm not sure what we would want to default to if there isn't an anything in `ancestry.interior`. I feel like we should error here, but the way I'm familiar with doing error handling in rust would require me to change the fn signature to be `Result<(), Error>` and I haven't seen that elsewhere in this file. How does Result<(), ()> work? From my perspective, this gives you no information on how something failed, right?

Reply:
> will ancestry ever have parents

No, as I have mentioned in the issue, `ancestry` is not supposed to have parents, and in fact in an upcoming XCMv3 PR, we have specifically typed it as `InteriorMultiLocation`, a change from just `MultiLocation`.

The default of using `OnlyChild` looks fine to me.